modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java

```diff
@@ -0,0 +1,531 @@
+ * and take in any other necessary values based on the arguments passed into the
+ * lambda function/reference method; it will then make a delegated call to the
+ * actual lambda function/reference method
+ *
+ * Take for example the following Painless script:
+ *
+ * {@code
+ * List list1 = new ArrayList(); "
+ * list1.add(2); "
+ * List1 list2 = new ArrayList(); "
+ * list1.forEach(x -> list2.add(x));"
+ * return list[0]"
+ * }
+ *
+ * The script contains a lambda function with a captured variable.
+ * The following Lambda class would be generated:
+ *
+ * {@code
+ *     public static final class $$Lambda0 implements Consumer {
+ *         private List arg$0;
+ *
+ *         public $$Lambda0(List arg$0) {
+ *             this.arg$0 = arg$0;
+ *         }
+ *
+ *         public static $$Lambda0 get$Lambda(List arg$0) {
+ *             return $$Lambda0(arg$0);
+ *         }
+ *
+ *         public void accept(Object val$0) {
+ *             Painless$Script.lambda$0(this.arg$0, val$0);
+ *         }
+ *     }
+ *
+ *     public class Painless$Script implements ... {
+ *         ...
+ *         public static lambda$0(List list2, Object x) {
+ *             list2.add(x);
+ *         }
+ *         ...
+ *     }
+ * }
+ *
+ * Note if the above didn't have a captured variable then
+ * the factory method get$Lambda would not have been generated.
+ * Also the accept method actually uses an invokedynamic
+ * instruction to call the lambda$0 method so that
+ * {@link MethodHandle#asType} can be used to do the necessary
+ * conversions between argument types without having to hard
+ * code them.
+ *
+ * When the {@link CallSite} is linked the linked method depends
+ * on whether or not there are captures.  If there are no captures
+ * the same instance of the generated lambda class will be
+ * returned each time by the factory method as there are no
+ * changing values other than the arguments.  If there are
+ * captures a new instance of the generated lambda class will
+ * be returned each time with the captures passed into the
+ * factory method to be stored in the member fields.
+ */
+public final class LambdaBootstrap {
+
+    /**
+     * Metadata for a captured variable used during code generation.
+     */
+    private static final class Capture {
+        private final String name;
+        private final Type type;
+        private final String desc;
+
+        /**
+         * Converts incoming parameters into the name, type, and
+         * descriptor for the captured argument.
+         * @param count The captured argument count
+         * @param type The class type of the captured argument
+         */
+        private Capture(int count, Class<?> type) {
+            this.name = "arg$" + count;
+            this.type = Type.getType(type);
+            this.desc = this.type.getDescriptor();
+        }
+    }
+
+    /**
+     * A counter used to generate a unique name
+     * for each lambda function/reference class.
```

Comment:
It's not worth complicating this to have to reload this class on a per-script basis.  If you really want know the lambda count during debugging you can run one script at a time in a debugger.  Plus, you also already know the answer since you generate exactly one lambda factory method per lambda.